### DISCUSSION TITLE
Autosuggest suggestion does not match actual completion

### DISCUSSION THREAD

**Discussion by fiso64** [chrisant996/clink#710] [Q&A] [answered]
This might be a bug, I'm not sure if I'm misunderstanding. I have autosuggest strategy set to `completion`. I expect the completion I get when pressing tab to be the same as the grey text, but sometimes they're not the same:

![image](https://github.com/user-attachments/assets/a05d14e6-2024-415f-940d-b73668584fc0)
![image](https://github.com/user-attachments/assets/e8c276aa-59e7-4046-8ee6-d0946e7fdfd0)

**Comment by chrisant996** [comment]
Those are two different features -- "auto-suggestions" (gray text) versus "completions".

Refer to docs at [How Completion Works](https://chrisant996.github.io/clink/clink.html#how-completion-works), which starts by describing the difference between the two features, and also has a link to the documentation section for the auto-suggestions feature.

Also:
- This is duplicate of #637, #475, #440, and several other issues.
- This is closely related to #556, #540, #529, and several other issues.

**Reply by fiso64** [reply]
I read the docs, and I know that in general suggestions are not the same as completions. What's confusing me is that they're not the same even when the autosuggest strategy is set to `completion`. Can you explain a little more what it does exactly, and why it's different from the first available completion?

**Reply by chrisant996** [reply]
I see.  Here is more information:

Inserting a suggestion never changes what's already been typed (except to match case, per the [autosuggest.original_case](https://chrisant996.github.io/clink/clink.html#autosuggest_original_case) setting).

So, if you type `m` but there's a `.mzzz` file, it cannot suggest it because `m` was typed, not `.m`.

But if you have <kbd>Tab</kbd> configured to use Windows style completion and cycle through matches, then the Windows style completion behavior also means `.mzzz` comes before `maaa` (try it in cmd.exe without Clink).

If you want bash-like completion behavior with respect to wildcards and leading dot, then turn off the [match wild](https://chrisant996.github.io/clink/clink.html#match_wild) setting.

**Reply by fiso64** [reply]
I set `match.wild = false`, but the behavior is still the same for the completion, i.e it completes to `.matplotlib` when I press tab after typing `m` without a leading dot.

**Reply by chrisant996** [reply]
Sorry, I misremembered which setting, and I check the documentation instead of the source code.  The documentation is inaccurate on this point.  I'll fix the documentation.

The leading dot behavior is tied to the [clink.default_bindings](https://chrisant996.github.io/clink/clink.html#default_bindings) setting.  That setting also controls default key bindings, so changing it will also affect the default key bindings for the keys listed in its documentation.

The reason the leading dot behavior is tied to that setting is so a user can change a single setting to switch between "Windows-like" versus "bash-like" completion experience.  Completing leading dots even when not typed is a standard Windows completion behavior.

If you switch the clink.default_bindings setting to `bash` then you may also need to make some changes in your `.inputrc` file to override any of those key bindings that you prefer to behave like Windows does (for example, to keep the Windows-like behavior where <kbd>Tab</kbd> cycles through completions).

**Reply by chrisant996** [reply]
I found the other way to control leading dots behavior:

You can set the [match-hidden-files](https://chrisant996.github.io/clink/clink.html#configmatchhiddenfiles) config variable to `off` in your `.inputrc` file.  That would also stop matching leading dots unless actually typed, and without any side effects on key bindings.

**Reply by fiso64** [reply]
that did the trick, thanks!

**Reply by fiso64** [reply]
I found a little issue: Apparently, 'hidden files' are both files that start with a `.`, as well as files that are marked as 'hidden' in windows. This means that with this variable set to off, it will never match your AppData folder.

**Reply by fiso64** [reply]
Turns out this doesn't resolve the issue even in the case of non-hidden files. In my case, typing `cd AppData\Local\Pack` will show an autosuggestion for `AppData\Local\Packages`, but completes to `AppData\Local\Package Cache`. Both strings start with `Pack`, so I'm not sure why that is.

**Reply by chrisant996** [reply]
I'm sorry that the auto suggestion feature is confusing. 

The additional things you've observed are correct behaviors.

It might be best to turn off the auto suggestions feature, to eliminate further confusion.

**Reply by fiso64** [reply]
In that case, please consider updating the documentation. I can't find anything that would imply they can be different with strategy set to `completion`. Under `autosuggest.strategy`, the docs simply state: 
> completion chooses the first of the matching completions

which given the above doesn't seem to be correct.

**Reply by chrisant996** [reply]
Please suggest a phrasing that would clarify the documentation for you.

- Do you want it to say "completion chooses the first completion that matches what was typed"?  How is that different than "first of the matching completions", apart from being longer?  The "menu complete" family of completion commands are intentionally not restricted to matching only what has been literally typed, but the auto-suggestions feature _**is**_ restricted to matching only what has been literally typed (because it can only show a preview of what hasn't been typed yet; there's no way to "show" how already-typed text would change).
- Do you want it to mention all the settings and input combinations and key bindings where it won't behave the way that you're requesting?  That requires multiple paragraphs, and inserting them there would create more confusion than it solves.

And we haven't yet gotten into how [match.ignore_case](https://chrisant996.github.io/clink/clink.html#match_ignore_case) or [match.ignore_accent](https://chrisant996.github.io/clink/clink.html#match_ignore_accent) affect the issue...

I understand the desire to try to change the auto-suggest feature into merely being an exact preview of "what TAB will do, if and only if TAB is bound to one of the `menu-complete` commands" (since that's the only kind of completion that cycles through matches).

I strongly recommend to instead turn off the auto-suggestion feature, because that's not how the feature is designed and using it that way loses almost all of its capabilities, while also breaking a bunch of subtle Windows standard completion behaviors (and I've already received feedback from accidentally breaking them before, so I'm not going to intentionally break them).

Press <kbd>Right</kbd> _**not**_ <kbd>Tab</kbd> to insert an auto-suggestion.

Pressing <kbd>Tab</kbd> is a completely different feature than auto-suggestion (pun intended), and the only way it can be hacked to seem similar to auto-suggestions is in a degenerate case where TAB is bound to `menu-complete` and autosuggest.strategy contains only `completion`.  In any other configuration, it's very clear that the two features are not the same at all.

And this is why I am leaning towards turning auto-completion OFF by default (see the poll in #644 -- "Should Auto-Suggestions be OFF by default?").  It causes too much confusion for it to be on by default.

I cannot make the auto-suggestions feature work how you want, without breaking other standard completion behaviors.  And I cannot figure out how to update the documentation without inserting multiple paragraphs.

You'll have to help me.

**Reply by fiso64** [reply]
I don't know how to describe this in the documentation because I have no idea what it actually does. The completion differing from the suggestion when it doesn't start with the same substring is predictable, I guess. But this doesn't fully explain the differences. Again, `AppData\Local\Pack` will show a suggestion for `AppData\Local\Packages`, but completes to `AppData\Local\Package Cache`. So it's really not the "first of the matching completions" no matter how you look at it. 

By the way, I actually had autosuggestions set to off for a long time. Then I noticed that there was (seemingly) a mode in which they would be the same as the completions, but which turned out to be wrong. So it would have made no difference whether it was on or off by default in my case. I guess what could have helped is a little note about the fact that even when the strategy is set to 'completion', the suggestions and actual completions may differ.

**Reply by chrisant996** [reply]
> But this doesn't fully explain the differences. Again, `AppData\Local\Pack` will show a suggestion for `AppData\Local\Packages`, but completes to `AppData\Local\Package Cache`. So it's really not the "first of the matching completions" no matter how you look at it. 

`AppData\Local\Package Cache` requires quotes.

You said it completes to `AppData\Local\Package Cache`, but no, it completes to `"AppData\Local\Package Cache"`.

Since you didn't type a beginning quote, autosuggestions cannot show a preview of inserting a quote into preceding text you already typed earlier.  And so it can't really choose that as a suggestion _that matches what's been typed so far_, because it **doesn't** match what's been typed so far.


> I guess what could have helped is a little note about the fact that even when the strategy is set to 'completion', the suggestions and actual completions may differ. 

Cool.  I can come up with something that explains that there will be differences, without going into detail about all the specific cases.  I may be able to find someplace in the standalone documentation (the html docs) to mention examples to show the principle of _why_ differences exist, without necessarily listing all the specific combinations and possibilities.

**Reply by fiso64** [reply]
>Since you didn't type a beginning quote, autosuggestions cannot show a preview of inserting a quote into preceding text you already typed earlier

Ohh, of course! I should have thought about this a little longer.

Is it true that all differences can be explained by this, i.e the fact that sometimes it would have to modify the text that has already been typed? 
If so, here's an alternative that I think might be a better 'completion' strategy:

Instead of showing the first matching completion which doesn't require inserting any text, really show the first completion, but ignore everything that would have to be inserted. For example, `AppData\Local\Pack` will show a suggestion for `AppData\Local\Package Cache"` (without a quote on the left), and if the user has a hidden directory called `.hidden-dir` and types `hidde` it will show `hidden-dir`, without the leading dot.

There is obviously still a discrepancy between completions and the shown suggestions, but it's closer, and I suspect that users who choose the 'completion' strategy would generally like them to be as close as possible. I think this behavior would be more intuitive. However, the insert suggestion command would need to be modified to insert text in what has already been typed when necessary.

Would that work?

**Reply by chrisant996** [reply, marked as answer]
> Is it true that all differences can be explained by this, i.e the fact that sometimes it would have to modify the text that has already been typed? 

Yes.  But there are multiple variations of ways.

> If so, here's an alternative that I think might be a better 'completion' strategy:
> 
> Instead of showing the first matching completion which doesn't require inserting any text, really show the first completion, but ignore everything that would have to be inserted. For example, `AppData\Local\Pack` will show a suggestion for `AppData\Local\Package Cache"` (without a quote on the left), and if the user has a hidden directory called `.hidden-dir` and types `hidde` it will show `hidden-dir`, without the leading dot.
> 
> There is obviously still a discrepancy between completions and the shown suggestions, but it's closer, and I suspect that users who choose the 'completion' strategy would generally like them to be as close as possible. I think this behavior would be more intuitive. However, the insert suggestion command would need to be modified to insert text in what has already been typed when necessary.
> 
> Would that work? 

No.  There are many, many problems with that.

I've spent tens of hours doing research and prototypes.  The ideas you're suggesting aren't new.  I've considered and explored all of these and found the many ways and cases where they just don't work.

If you want to pursue that, you can.  For me, though, this is just rehashing stuff I've already researched deeply.  I'm not going to engage in this further.  You're overlooking many things, and as you said, you need to think about the details longer and more deeply.

**Reply by fiso64** [reply]
Understood. I'll try to see if I can implement this naive approach as a custom suggester script for personal use.
